File: android/digia-engage/src/main/java/com/digia/engage/internal/CampaignStore.kt
package com.digia.engage.internal

import com.digia.engage.internal.model.CampaignModel
import java.util.concurrent.ConcurrentHashMap

internal class CampaignStore {
    private val campaigns = ConcurrentHashMap<String, CampaignModel>()

    fun populate(list: List<CampaignModel>) {
        campaigns.clear()
        list.forEach { campaigns[it.campaignKey] = it }
    }

    fun find(campaignKey: String): CampaignModel? = campaigns[campaignKey]

    fun findById(campaignId: String): CampaignModel? =
        campaigns.values.firstOrNull { it.id == campaignId }

    fun isEmpty(): Boolean = campaigns.isEmpty()
}

File: flutter/lib/api/internal/campaign/campaign_store.dart
import 'campaign_model.dart';

/// In-memory cache of campaigns keyed by `campaignKey`.
///
/// Dart port of the Android `CampaignStore`. Populated once during
/// [DigiaInstance.initialize] and consulted on every CEP-triggered campaign.
/// Not persisted across launches — the SDK refetches on each cold start.
class CampaignStore {
  final Map<String, CampaignModel> _campaigns = <String, CampaignModel>{};

  /// Replaces the entire cache with [list], keyed by `campaignKey`.
  void populate(List<CampaignModel> list) {
    _campaigns.clear();
    for (final campaign in list) {
      _campaigns[campaign.campaignKey] = campaign;
    }
  }

  CampaignModel? find(String campaignKey) => _campaigns[campaignKey];

  CampaignModel? findById(String campaignId) {
    for (final campaign in _campaigns.values) {
      if (campaign.id == campaignId) return campaign;
    }
    return null;
  }

  bool get isEmpty => _campaigns.isEmpty;

  void clear() => _campaigns.clear();
}
